query_executor.go

```diff
@@ -50,51 +121,46 @@ func (q *queryExecutor) executeQuery(qry ExecutableQuery) (*Iter, error) {
 			continue
 		}
 
-		iter = q.attemptQuery(qry, conn)
-		// Update host
-		hostResponse.Mark(iter.err)
-
-		if rt == nil {
-			iter.host = host
-			break
-		}
-
-		switch rt.GetRetryType(iter.err) {
-		case Retry:
-			for rt.Attempt(qry) {
-				iter = q.attemptQuery(qry, conn)
-				hostResponse.Mark(iter.err)
-				if iter.err == nil {
-					iter.host = host
-					return iter, nil
-				}
-				if rt.GetRetryType(iter.err) != Retry {
-					break
-				}
-			}
-		case Rethrow:
-			return nil, iter.err
-		case Ignore:
-			return iter, nil
-		case RetryNextHost:
+		select {
+		case <-stop:
+			// stop this execution and return
+			return
 		default:
-		}
+			// Run the query
+			iter = q.attemptQuery(qry, conn)
+			iter.host = selectedHost.Info()
+			// Update host
+			selectedHost.Mark(iter.err)
+
+			// Exit if the query was successful
+			// or no retry policy defined or retry attempts were reached
+			if rt == nil || iter.err == nil || !rt.Attempt(qry) {
```

Comment:
Reorder the statements according to the comment (which is also the logical order to check, even though, yes, it doesn't really matter – a nit is a nit is a nit):
```go
if iter.err == nil || rt == nil || !rt.Attempt(qry) {
```

Reply:
Right, came back after I read your comment over in the [other PR](https://github.com/gocql/gocql/pull/1159/files#r218965827). I think we should take another stab at pulling #1151, #1164, and #1162 together into a coherent whole so that we can put speculative on top. I'm not harking on your implementation in here – it reads really fluidly. What it doesn't do right now is handle the case if non-idempotent query being executing speculatively and we could do that if #1162 were implemented by explicitly stopping retries in those cases like write errors as well. On the other hand… one has to specifically turn on speculative execution on a query, right? 🤔 I may be overthinkingn this in that case.

Reply:
@annismckenzie Yea, I was thinking to myself whether you're getting back to this :-) So indeed, it is possible that non-idempotent query will be retried, but definitely not in a speculative execution. `Speculative` in this case means that there are multiple executions running at the same time, and the first one to get a response wins. We take care of that in the entrance to the `executeQuery` by forcing the speculative execution policy to NonSpeculative if the policy is not idempotent.

Back to the issue. I'm all for fixing the retries to ack the idempotence, and definitely would be for taking another stab at error handling. I'm not overly sure these depend on this work, they can be layered in any order. It can be pretty decent in size, especially if we include the retries/idempotence handling and actually wrapping errors in their own types. Do you need a hand with that one?

Reply:
I would really, really, really appreciate that! 🤝 I'll text you via other channels.

Reply:
:+1: